src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -27251,6 +27363,128 @@ static void M1<T>(T t1)
+{
+    [AllowNull, NotNull]public TClass P2 { get; set; } = null!;
+    [AllowNull, NotNull]public TClass? P3 { get; set; } = null;
+}
+public class CStruct<TStruct> where TStruct : struct
+{
+    [AllowNull, NotNull]public TStruct? P5 { get; set; }
+}
+class Program
+{
+    static void M1<T>([MaybeNull]T t1)
+    {
+        var xOpen = new COpen<T>();
+        xOpen.P1 = t1;
+        xOpen.P1.ToString();
+    }
+    static void M2<T>(T t2) where T : class
+    {
+        var xOpen = new COpen<T>();
+        xOpen.P1 = null;
+        xOpen.P1.ToString();
+
+        var xClass = new CClass<T>();
+        xClass.P2 = null;
+        xClass.P2.ToString();
+        xClass.P3 = null;
+        xClass.P3.ToString();
+    }
+    static void M5<T>(T? t5) where T : struct
+    {
+        var xOpen = new COpen<T?>();
+        xOpen.P1 = null;
+        xOpen.P1.ToString();
+
+        var xStruct = new CStruct<T>();
+        xStruct.P5 = null;
+        xStruct.P5.Value.ToString();
+    }
+}";
+            var comp = CreateNullableCompilation(new[] { source, AllowNullAttributeDefinition, NotNullAttributeDefinition, MaybeNullAttributeDefinition });
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void AllowNull_Property_InCompoundAssignment()
+        {
+            var lib_cs =
+@"using System.Diagnostics.CodeAnalysis;
+public class C
+{
+    [AllowNull] public C P { get; set; }
+    public static C? operator +(C? x, C? y) => throw null!;
+}";
+            var lib = CreateNullableCompilation(new[] { lib_cs, AllowNullAttributeDefinition });
+            lib.VerifyDiagnostics(
+                // (4,26): warning CS8618: Non-nullable property 'P' is uninitialized. Consider declaring the property as nullable.
+                //     [AllowNull] public C P { get; set; }
+                Diagnostic(ErrorCode.WRN_UninitializedNonNullableField, "P").WithArguments("property", "P").WithLocation(4, 26)
+                );
+
+            var source = @"
+class Program
+{
+    static void M(C c)
+    {
+        c.P += null;
+        c.P.ToString();
+    }
+}";
+
+            var comp = CreateNullableCompilation(source, references: new[] { lib.EmitToImageReference() });
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void AllowNull_Property_InDeconstructionAssignment()
+        {
+            var source =
+@"using System.Diagnostics.CodeAnalysis;
+public class C
+{
+    [AllowNull] public C P { get; set; } = null;
+}
+class Program
+{
+    void M(C c)
+    {
+        (c.P, _) = (null, 1);
+        c.P.ToString();
+
+        ((c.P, _), _) = ((null, 1), 2);
+        c.P.ToString();
+
+        (c.P, _) = this;
+        c.P.ToString();
+
+        ((_, c.P), _) = (this, 1);
+        c.P.ToString();
+    }
+    void Deconstruct(out C? x, out C? y) => throw null!;
+}";
+            // Note: we warn on the property initializer because it is a direct assignment to the backing field
```

Comment:
what is the expected behavior for `[AllowNull, MaybeNull] C Prop { get; set; } = null;`?